test/jobs/trace_importer_job_test.rb

```diff
@@ -0,0 +1,64 @@
+require "test_helper"
+require "minitest/mock"
+
+class TraceImporterJobTest < ActiveJob::TestCase
+  def test_success_notification
+    # Check that the user gets a success notification when the trace has valid points
+    trace = create(:trace)
+
+    gpx = Minitest::Mock.new
+    def gpx.actual_points
+      5
+    end
+
+    trace.stub(:import, gpx) do
+      perform_enqueued_jobs do
+        TraceImporterJob.perform_now(trace)
+      end
+    end
```

Comment:
I'm struggling to understand this, or to find any documentation of the `stub` method that would help me. I assume it's a FactoryBot thing but although I find occasional references to it on google the actual documentation never seems to mention it.

I assume it stubs the `import` method but I don't see when this test would ever call that method? or why calling import which would normally be called from the job should actually run the job?

I'm probably just being dense but I'd like to understand what's going on ;-)

Reply:
It's from minitest/mock. 

https://www.rubydoc.info/gems/minitest/4.2.0/Object:stub

The stubbing takes a block, and the stub is only effective inside the block. So when we call `TraceImporterJob.perform_now(trace)`, the `trace.import` method has been overwritten to just return the (fake) gpx object and avoid doing anything else (parsing xml, creating images on disk etc).

I did all this before realising that I needed the FakeFS for other stuff anyway. So if you prefer, I could rework this to create a fully working `trace` object that actually runs `trace.import` for real, with suitable setup to put a valid xml file in the right place and clean up the side-effects of running the `.import` code. Or we can stick with mocking/stubbing, which is probably a better way to do it, but perhaps a bit impenetrable?